### PR TITLE
Convert some general.shard base tests to null safety

### PR DESCRIPTION
**Pull request by jmagman**
Also move `general.shard/base/logs_test.dart` to more appropriate `commands.shard/hermetic/logs_test.dart`.

Part of #71511

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/context.dart [hunk 1]
```diff
@@ -118,7 +118,7 @@ class AppContext {
     if (value == null && _parent != null) {
       value = _parent!.get<T>();
     }
-    return _unboxNull(value ?? _generateIfNecessary(T, _fallbacks)) as T;
+    return _unboxNull(value ?? _generateIfNecessary(T, _fallbacks)) as T?;
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/base/context.dart:121]: Ruh roh

File: packages/flutter_tools/test/general.shard/base/async_guard_test.dart [hunk 2]
```diff
@@ -288,7 +287,7 @@ void main() {
         if (!completer.isCompleted) {
           completer.complete(null);
         }
-      }));
+      })!);
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/base/async_guard_test.dart:290]: what are these ones for 🤔

File: packages/flutter_tools/test/general.shard/base/deferred_component_test.dart [hunk 3]
```diff
@@ -56,10 +53,10 @@ void main() {
       component.assignLoadingUnits(loadingUnits1);
 
       expect(component.assigned, true);
-      expect(component.loadingUnits.length, 2);
-      expect(component.loadingUnits.contains(loadingUnits1[0]), true);
-      expect(component.loadingUnits.contains(loadingUnits1[1]), true);
-      expect(component.loadingUnits.contains(loadingUnits1[2]), false);
+      expect(component.loadingUnits!.length, 2);
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/base/deferred_component_test.dart:56]: You could avoid the `!` by writing these as:

`expect(component.loadingUnits, contains(loadingUnits1[0]));`

**Review by jonahwilliams — APPROVED** [review]
LGTM with nits

File: packages/flutter_tools/test/general.shard/base/async_guard_test.dart (hunk 2 above)
Comment (jmagman) [reply packages/flutter_tools/test/general.shard/base/async_guard_test.dart:290]: `unawaited` takes a nonnull `Future` and `runZonedGuarded` can be null.

However a better change is to update out `unawaited` to `void unawaited(Future<void>? future)` and match the signature in https://pub.dev/documentation/pedantic/latest/pedantic/unawaited.html
https://github.com/flutter/flutter/blob/a2f67720ef146ec44a35800706128179c33e43a9/packages/flutter_tools/lib/src/base/common.dart#L36

Good eye.

File: packages/flutter_tools/test/general.shard/base/deferred_component_test.dart (hunk 3 above)
Comment (jmagman) [reply packages/flutter_tools/test/general.shard/base/deferred_component_test.dart]: And `hasLength`.  Done.

**Comment by jmagman** [comment]
Kicking the tree.

**Pull request merged**